cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemotePluginFinder.java

```diff
@@ -209,10 +217,74 @@ public Location getArtifactLocation(ArtifactId artifactId)
 
     String path = response.getResponseBodyAsString();
     Location location = Locations.getLocationFromAbsolutePath(locationFactory, path);
+
+    // If the artifact doesn't exist locally then fetch it from app fabric and save it locally
     if (!location.exists()) {
-      throw new IOException(String.format("Artifact Location does not exist %s for artifact %s version %s",
-                                          path, artifactId.getArtifact(), artifactId.getVersion()));
+      LOG.debug(String.format("Artifact '%s' is not present locally at %s, attempting to fetch it from app-fabric.",
+                              artifactId.getArtifact(), path));
+
+      Retries.runWithRetries(() -> getAndStoreArtifact(artifactId, location), retryStrategy);
     }
     return location;
   }
+
+  private void getAndStoreArtifact(ArtifactId artifactId, Location location) throws IOException {
+    String namespaceId = artifactId.getNamespace();
+    ArtifactScope scope = ArtifactScope.USER;
+    // Cant use 'system' as the namespace in the request because that generates an error, the namespace doesnt matter
+    // as long as it exists. Using default because it will always be there
+    if (ArtifactScope.SYSTEM.toString().equalsIgnoreCase(namespaceId)) {
```

Comment:
Why are we comparing scope to namespace?

Reply:
Scope and namespace are basically the same thing for the `system` namespace. We need to update the scope to `SYSTEM` if the given namespace is "system". We cant pass "system" as the namespace value because it throws a NamespaceNotFound error.

Reply:
I misunderstood Terence's comment, the `ArtifactScope.SYSTEM` should be `NamespaceId.SYSTEM`. It is fixed now.